arangod/Pregel/StatusWriter/CollectionStatusWriter.h

```diff
@@ -0,0 +1,191 @@
+    auto handleOperationResult =
+        [](SingleCollectionTransaction& trx, OperationOptions& options,
+           Result& transactionResult,
+           OperationResult&& opRes) -> OperationResult {
+      transactionResult = trx.finish(opRes.result);
+      if (transactionResult.fail() && opRes.ok()) {
+        return OperationResult{std::move(transactionResult), options};
+      }
+      return opRes;
+    };
+
+    AccessMode::Type accessModeType;
+    if (operation == OperationType::CREATE_DOCUMENT ||
+        operation == OperationType::UPDATE_DOCUMENT ||
+        operation == OperationType::DELETE_DOCUMENT) {
+      accessModeType = AccessMode::Type::WRITE;
+    } else {
+      TRI_ASSERT(operation == OperationType::READ_DOCUMENT);
+      accessModeType = AccessMode::Type::READ;
+    }
+
+    // transaction options
+    SingleCollectionTransaction trx(ctx(), StaticStrings::PregelCollection,
+                                    accessModeType);
+    trx.addHint(transaction::Hints::Hint::SINGLE_OPERATION);
+    OperationOptions options(ExecContext::current());
+
+    // begin transaction
+    Result transactionResult = trx.begin();
+    if (transactionResult.fail()) {
+      return OperationResult{std::move(transactionResult), options};
+    }
+
+    // execute transaction
+    auto payload = inspection::serializeWithErrorT(data);
+    switch (accessModeType) {
+      case AccessMode::Type::WRITE: {
+        if (operation == OperationType::UPDATE_DOCUMENT) {
+          return handleOperationResult(
+              trx, options, transactionResult,
+              trx.update(StaticStrings::PregelCollection, payload->slice(),
+                         {}));
+        } else if (operation == OperationType::CREATE_DOCUMENT) {
+          return handleOperationResult(
+              trx, options, transactionResult,
+              trx.insert(StaticStrings::PregelCollection, payload->slice(),
+                         {}));
+        } else {
+          TRI_ASSERT(operation == OperationType::DELETE_DOCUMENT);
+          return handleOperationResult(
+              trx, options, transactionResult,
+              trx.remove(StaticStrings::PregelCollection, payload->slice(),
+                         {}));
+        }
+      }
+      case AccessMode::Type::READ: {
+        // Note: documentAsync can throw.
+        return handleOperationResult(
+            trx, options, transactionResult,
+            trx.documentAsync(StaticStrings::PregelCollection, payload->slice(),
+                              {})
+                .get());
+      }
+      default: {
+        TRI_ASSERT(false);
+      }
+    }
+
+    TRI_ASSERT(false);
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_BAD_PARAMETER);
+  }
+
+  [[nodiscard]] std::shared_ptr<transaction::Context> ctx() const {
+    // we must use v8
+    return transaction::V8Context::CreateWhenRequired(_vocbase, true);
+  }
+
+ private:
+  TRI_vocbase_t& _vocbase;
```

Comment:
```suggestion
  TRI_vocbase_t& _vocbase;
```
We should use a DatabaseGuard here.